### ISSUE TITLE
QK^T dimension problem

### ISSUE THREAD

**Issue by siihwanpark** [psh150204/Transformer#1]
In the decoder block, Q is from the previous block and K is from the encoder.

This case, Q will be a tensor with size **[batch_size, src_num_of_words, d_model]** and K will be a tensor with size **[batch_size, trg_num_of_words, d_model]**. 

Thus, QK_transpose will be a tensor with size **[batch_size, src_num_of_words, trg_num_of_words]**.

If src_num_of_words != trg_num_of_words, then the mask is unavailable.

How to solve this issue?

```python
def scaled_dot_product_attention(Q, K, V, mask):
    # input format
    # Q : tensor with size [batch_size, num_of_words, d_k]
    # K : tensor with size [batch_size, num_of_words, d_k]
    # V : tensor with size [batch_size, num_of_words, d_v]
    # mask : tensor with size [batch_size, num_of_words, num_of_words]
    
    d_k = Q.size(2) # scale factor
    Kt = torch.transpose(K, -2, -1) # batch_size * d_k * num_of_words
    QKt = torch.bmm(Q,Kt) # batch_size * num_of_words * num_of_words

    if Q.size() != K.size():
        print(Q.size(), K.size())
    
    masked_QKt = QKt.masked_fill(mask == 1, -1e9)

    weights = F.softmax(masked_QKt/np.sqrt(d_k)) # batch_size * num_of_words * num_of_words
    return torch.bmm(weights, V) # batch_size * num_of_words * d_v
```

**Comment by siihwanpark** [comment]
make another mask which is called 'memory mask'.
```python
def generate_masks(src, trg, src_padding, trg_padding):
    # input
    # src : a tensor with size [batch_size, src_sentence_length]
    # trg : a tensor with size [batch_size, trg_sentence_length]

    src_n = src.size(1) # src_sentence_length
    trg_n = trg.size(1) # trg_sentence_length

    src_padding_mask = (src == src_padding).unsqueeze(1) # batch_size * 1 * src_sentence_length
    trg_padding_mask = (trg == trg_padding).unsqueeze(1) # batch_size * 1 * trg_sentence_length

    src_mask = src_padding_mask.repeat(1, src_n, 1) # batch_size * src_sentence_length * src_sentence_length
    memory_mask = src_padding_mask.repeat(1, trg_n, 1) # batch_size * trg_sentence_length * src_sentence_length
    trg_mask = trg_padding_mask.repeat(1, trg_n, 1) # batch_size * trg_sentence_length * trg_sentence_length

    return src_mask, memory_mask, trg_mask
```
